### PR TITLE
Configurable client for faster timeouts

### PR DESCRIPTION
**Pull request by marians**
This PR adds a way to configure the API client for

- quicker timeouts where appropriate, e. g. when listing organizations
- Setting a user agent string

This is applied for all API requests.

The motivation were lots of timeouts when using our current leaseweb-g8s backend, especially in commands like `list clusters`, where multiple requests are made.

Technically this could have been part of `gsclientgen`, but in fact I don't want to add code there, as all code in that repository is generated by swagger.

### CODE REVIEW

File: commands/list_clusters.go [hunk 1]
```diff
@@ -43,9 +44,14 @@ func checkListClusters(cmd *cobra.Command, args []string) error {
 
 // list all clusters the user has access to
 func listClusters(cmd *cobra.Command, args []string) {
-	client := gsclientgen.NewDefaultApiWithBasePath(cmdAPIEndpoint)
+	clientConfig := client.Configuration{
+		Endpoint:  cmdAPIEndpoint,
+		Timeout:   3 * time.Second,
```
Comment (xh3b4sd) [inline commands/list_clusters.go]: What is actually the reasoning behind the very different timeouts?

File: client/client_test.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	"strings"
+	"testing"
+	"time"
+)
+
+func TestTimeout(t *testing.T) {
+	clientConfig := Configuration{
+		Endpoint: "http://httpbin.org/delay/2?",
```
Comment (marians) [inline client/client_test.go]: I think depending on the network for tests should be fine. But I will look at the alternative. Having it totally internal is definitely better.

File: commands/list_clusters.go (hunk 1 above)
Comment (marians) [inline commands/list_clusters.go]: I want to fail fast in situations where multiple GET calls are made in sequence. E. g. when listing clusters, we first fetch all the organizations (memberships). Then we look for clusters in each org. This failing after one minute (as it happened just recently with flapping API behaviour) is just a really bad waste of time for the end user. I'd rather fail after a few seconds and let the user retry.

In cases of transactions (POST, PUT, PATCH) I'd like to tolerate longer timeouts, as we expect a little bit more work to happen on the other end, and retrying might not be as easy (unless transactions are implemented for that route).

The next step of course will be doing the retrying automatically, so the user doesn't have to bother at all in most cases.

It's a tiny step, but I see is as a little piece in a bigger resiliance picture.

File: client/client_test.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+		Timeout:  1 * time.Second,
+	}
+	apiClient := NewClient(clientConfig)
+	_, _, err := apiClient.GetUserOrganizations("foo", "foo", "foo", "foo")
+	if err == nil || !strings.Contains(err.Error(), "Client.Timeout exceeded") {
```
Comment (marians) [inline client/client_test.go]: I'll give that a try

Comment (marians) [inline client/client_test.go]: Unfortunately, the error type isn't specific enough: `*url.Error`.

File: commands/create_cluster.go [hunk 4]
```diff
@@ -335,8 +337,13 @@ func addCluster(cmd *cobra.Command, args []string) {
 			// command line flag overwrites
 			authHeader = "giantswarm " + cmdToken
 		}
-		client := gsclientgen.NewDefaultApiWithBasePath(cmdAPIEndpoint)
-		responseBody, apiResponse, _ := client.AddCluster(authHeader, addClusterBody, requestIDHeader, createClusterActivityName, cmdLine)
+		clientConfig := client.Configuration{
+			Endpoint:  cmdAPIEndpoint,
+			Timeout:   60 * time.Second,
```
Comment (marians) [inline commands/create_cluster.go]: Done

**Comment by marians** [comment]
@xh3b4sd Thanks for the review! Happy for a re-check.

File: client/client_test.go (hunk 3 above)
Comment (xh3b4sd) [inline client/client_test.go]: When it is a URL error, we can check if it is an timeout error. See https://golang.org/pkg/net/url/#Error.Timeout.

File: client/client_test.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+	"time"
+)
+
+func TestTimeout(t *testing.T) {
+	// Our test server.
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// enforce a timeout longer than the client's
+		time.Sleep(3 * time.Second)
+		fmt.Fprintln(w, "Hello")
+	}))
```
Comment (xh3b4sd) [inline client/client_test.go]: I like that way more than before. Cool.

**Comment by marians** [comment]
Changed error type checking.

**Pull request merged**